src/main/scala/sbtavro/SbtAvro.scala

```diff
@@ -49,11 +49,26 @@ object SbtAvro extends AutoPlugin {
 
     lazy val avroArtifactTasks: Seq[TaskKey[File]] = Seq(Compile, Test).map(_ / packageAvro)
 
+    val avroSchemaParser = taskKey[() => Schema.Parser](".avsc schema parser")
```

Comment:
Do we really need to have `TaskKey[() => Schema.Parser]` instead of `TaskKey[Schema.Parser]` as signature ?
I felt that tasks are executed every time if not cached.

Reply:
yes, this is unfortunate. The behaviour of Schema Parser is as such that it changes internal state while parsing new files - see `AvscFilesCompiler.java`. That is the reason why it is needed to actually re-create (or clone, but no such option is given) new Parser object and we can not reuse original from the task.

Reply:
I'd like to get rid of this

```
val schemaParser = new AtomicReference(new Schema.Parser())
```

which is just here to store the parser configuration.

What about using a builder setting `avroSchemaParserBuilder: SettingKey[SchemaParserBuilder]` with

```scala
public class SchemaParserBuilder {

    private final Schema.Parser parser = new Schema.Parser();

    private SchemaParserBuilder() { }

    public Schema.Parser build() {
        Schema.Parser copy = new Schema.Parser();
        copy.addTypes(parser.getTypes());
        copy.setValidate(parser.getValidate());
        copy.setValidateDefaults(parser.getValidateDefaults());
        return copy;
    }

    public SchemaParserBuilder withTypes(Map<String, Schema> types) {
        parser.addTypes(types);
        return this;
    }

    public SchemaParserBuilder withValidate(boolean validate) {
        parser.setValidate(validate);
        return this;
    }

    public SchemaParserBuilder withValidateDefaults(boolean validate) {
        parser.setValidateDefaults(validate);
        return this;
    }

    public static SchemaParserBuilder newBuilder() {
        return new SchemaParserBuilder();
    }
}
```

We can also define the extra `avroExternalTypes: SettingKey[Map[String, Schema]]` to ease external schema injection.

If user want to have a complete custom parser, they will have to give a new builder implementation.
Overall it's more code but it guides more the users.

Reply:
This PR is on a way to remove `val schemaParser = new AtomicReference(new Schema.Parser())` (annotating as deprecated).
I did not remove it because not wanting to break suddenly user code.
Builder might be a better api than supplier.
I think it might sense to deprecate settings `avroValidate` and `avroValidateDefaults`.

Then from user POV code would look like:
```scala
  avroSchemaParserBuilder := SchemaParserBuilder.newBuilder().withValidate(???). withValidateDefaults(???)
```

Reply:
Right. Let's avoid sudden breaking changes.
See github.com/sisidra/sbt-avro/pull/1